### PR TITLE
Fix bug in semantic parser when sliced array is modified in function call

### LINKED ISSUE #803: Bug in semantic parser when sliced array is modified in function call
**Describe the bug**

Pull request #716 introduced a regression bug at the semantic stage, which appears when these conditions are met:
- One function modifies an input array (which has `intent(inout)`);
- Another function calls the first function passing a sliced array to it;
- Both functions are pyccelized.

**To Reproduce**

Let's consider file `test_view.py`:

```python
import numpy as np

def f(x : 'float[:]'):
    x[0] = 1

def g():
    x = np.zeros(3)
    f(x[:])
    print(x)
```

Pyccel fails during the semantic stage:
```bash/fortran/c
$ pyccel test_view.py 
Traceback (most recent call last):
  File "/home/yamanguc/.local/bin/pyccel", line 11, in <module>
    load_entry_point('pyccel', 'console_scripts', 'pyccel')()
  File "/home/yamanguc/Library/PYCCEL/pyccel/pyccel/commands/console.py", line 258, in pyccel
    folder        = args.output)
  File "/home/yamanguc/Library/PYCCEL/pyccel/pyccel/codegen/pipeline.py", line 247, in execute_pyccel
    parser.annotate(**settings)
  File "/home/yamanguc/Library/PYCCEL/pyccel/pyccel/parser/parser.py", line 191, in annotate
    **settings)
  File "/home/yamanguc/Library/PYCCEL/pyccel/pyccel/parser/semantic.py", line 173, in __init__
    self.annotate()
  File "/home/yamanguc/Library/PYCCEL/pyccel/pyccel/parser/semantic.py", line 209, in annotate
    ast = self._visit(ast, **settings)
  File "/home/yamanguc/Library/PYCCEL/pyccel/pyccel/parser/semantic.py", line 802, in _visit
    obj = getattr(self, annotation_method)(expr, **settings)
  File "/home/yamanguc/Library/PYCCEL/pyccel/pyccel/parser/semantic.py", line 838, in _visit_CodeBlock
    ls = [self._visit(i, **settings) for i in expr.body]
  File "/home/yamanguc/Library/PYCCEL/pyccel/pyccel/parser/semantic.py", line 838, in <listcomp>
    ls = [self._visit(i, **settings) for i in expr.body]
  File "/home/yamanguc/Library/PYCCEL/pyccel/pyccel/parser/semantic.py", line 802, in _visit
    obj = getattr(self, annotation_method)(expr, **settings)
  File "/home/yamanguc/Library/PYCCEL/pyccel/pyccel/parser/semantic.py", line 2593, in _visit_FunctionDef
    assigned     = [a.name for a in all_assigned if a.rank > 0]
  File "/home/yamanguc/Library/PYCCEL/pyccel/pyccel/parser/semantic.py", line 2593, in <listcomp>
    assigned     = [a.name for a in all_assigned if a.rank > 0]
AttributeError: 'IndexedElement' object has no attribute 'name'

```

**Expected behavior**

Pyccel should be able to process file `test_view.py` without any problems, and a call to `g()` should print `[1, 0, 0]`.

**Language**

This bug is independent of the target language, as it happens during the semantic stage.

### PR DESCRIPTION
**Pull request by EmilyBourne**
Collect all `Variable` objects from assigned objects to fix #803 

**Commit Summary**

- Add test in a module;
- Remove obsolete `get_assigned_symbols` function;
- Add `__str__` method to `IndexedElement`;
- Replace call to `get_assigned_symbols` with 2 calls to `get_attribute_nodes` and 2 list comprehensions;
- Collect `Variable` objects from all assigned objects (fixes #803);
- [BUGFIX] Check function name not class name when assembling `apps` variable;
- Update patch number: new library version is 1.2.1.

### CODE REVIEW

File: tests/epyccel/test_epyccel_modules.py [hunk 1]
```diff
@@ -68,3 +68,23 @@ def test_module_3(language):
...
+    y = np.random.random_sample(n_y)
+
+    x_pyc = x.copy()
+    y_pyc = y.copy()
+
+    max_pyt = mod.f(x,y)
+    max_pyc = modnew.f(x_pyc, y_pyc)
+    assert np.isclose( max_pyt, max_pyc, rtol=1e-14, atol=1e-14 )
+    assert np.isclose( x, x_pyc, rtol=1e-14, atol=1e-14 ).all()
+    assert np.isclose( y, y_pyc, rtol=1e-14, atol=1e-14 ).all()
```
Comment (yguclu) [inline tests/epyccel/test_epyccel_modules.py:90]: Can't we use `np.allclose` here? It should be a bit faster.

File: tests/epyccel/modules/Module_6.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+            else:
+                i+=1
+    return x[-1]
+
+def f(x : 'float[:]', y : 'float[:]'):
+    n = x.shape[0]
+    for i in range(n):
+        set_i(x, i, float(i))
+    b = inplace_max(y)
+    return b
```
Comment (yguclu) [inline tests/epyccel/modules/Module_6.py:28]: Aren't we missing a test where we pass a **sliced** array to a function that modifies it?

**Review by yguclu — CHANGES_REQUESTED** [review]
Good job! We should update the library version to `1.2.1`

File: tests/epyccel/modules/Module_6.py (hunk 2 above)
Comment (EmilyBourne) [reply tests/epyccel/modules/Module_6.py:28]: Oops ^^

File: tests/epyccel/test_epyccel_modules.py (hunk 1 above)
Comment (EmilyBourne) [reply tests/epyccel/test_epyccel_modules.py]: Fixed in 69b6ee2

File: tests/epyccel/modules/Module_6.py (hunk 2 above)
Comment (EmilyBourne) [reply tests/epyccel/modules/Module_6.py]: Fixed in d1ffcdb

File: tests/epyccel/test_epyccel_modules.py (hunk 1 above)
Comment (yguclu) [reply tests/epyccel/test_epyccel_modules.py]: Thanks

**Review by yguclu — APPROVED** [review]

**Pull request merged**